extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/data/input/kafkainput/KafkaInputReader.java

```diff
@@ -0,0 +1,144 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+public class KafkaInputReader implements InputEntityReader
+{
+  private static final Logger log = new Logger(KafkaInputReader.class);
+  private final InputRowSchema inputRowSchema;
+  private final KafkaRecordEntity record;
+  private final KafkaHeaderReader headerParser;
+  private final InputEntityReader keyParser;
+  private final InputEntityReader payloadParser;
+  private final String keyLabelPrefix;
+  private final String recordTimestampLabelPrefix;
+
+  public KafkaInputReader(
+      InputRowSchema inputRowSchema,
+      KafkaRecordEntity record,
+      KafkaHeaderReader headerParser,
+      InputEntityReader keyParser,
+      InputEntityReader payloadParser,
+      String keyLabelPrefix,
+      String recordTimestampLabelPrefix
+  )
+  {
+    this.inputRowSchema = inputRowSchema;
+    this.record = record;
+    this.headerParser = headerParser;
+    this.keyParser = keyParser;
+    this.payloadParser = payloadParser;
+    this.keyLabelPrefix = keyLabelPrefix;
+    this.recordTimestampLabelPrefix = recordTimestampLabelPrefix;
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read() throws IOException
+  {
+    Map<String, Object> mergeList = new HashMap<>(this.headerParser.read());
+
+    if (this.keyParser != null) {
+      CloseableIterator<InputRow> keyIterator = this.keyParser.read();
+      // Key currently only takes the first row and ignores the rest.
+      if (keyIterator.hasNext()) {
+        MapBasedInputRow keyRow =  (MapBasedInputRow) keyIterator.next();
+        mergeList.put(
+            this.keyLabelPrefix + "key",
+            keyRow.getEvent().entrySet().stream().findFirst().get().getValue()
+        );
+      }
+      keyIterator.close();
+    }
+    // Add kafka record timestamp to the mergelist
+    mergeList.put(this.recordTimestampLabelPrefix + "timestamp", this.record.getRecord().timestamp());
+    //log.warn("LOKI Kafka record :" + new String(this.record.getRecord().value(), StandardCharsets.UTF_8));
+
+    CloseableIterator<InputRow> iterator = this.payloadParser.read();
+    List<InputRow> rows = new ArrayList<>();
+    try {
+      while (iterator.hasNext()) {
+      /* Currently we prefer payload attributes if there is a collision in names.
+          We can change this beahvior in later changes with a config knob. This default
+          behavior lets easy porting of existing inputFormats to the new one without any changes.
+       */
+        MapBasedInputRow row = (MapBasedInputRow) iterator.next();
+        Map<String, Object> event = new HashMap<>(row.getEvent());
+        mergeList.forEach((key, value) -> event.merge(key, value, (v1, v2) -> v1));
+
+        HashSet<String> newDimensions = new HashSet<String>(row.getDimensions());
+        newDimensions.addAll(mergeList.keySet());
+        // Free mergeList
+        mergeList = null;
```

Comment:
won't this cause the line above to throw a null-pointer exception on the next row?